v2/apiserver/internal/core/mongodb/projects_store.go

```diff
@@ -120,15 +121,81 @@ func (p *projectsStore) ListSubscribers(
 	event core.Event,
 ) (core.ProjectList, error) {
 	projects := core.ProjectList{}
-	subscriptionMatchCriteria := bson.M{
+	// Finding all Projects that are subscribed to a given Event is rather tricky.
+	// Matching on the basis of source, type, and qualifiers is easy enough, but
+	// LABELS account for the difficulty. If we were selecting Events, it's easy
+	// enough to say "select all events that have (at least) these labels," but
+	// when selecting PROJECTS, we're effectively working in reverse. We want to
+	// select all Projects having an EventSubscription that matches the Event's
+	// source, type, and qualifiers, AND doesn't have any labels that PRECLUDE a
+	// match. The only practical way of expressing those criteria in a MongoDB
+	// query is to provide a JavaScript function as a "where" clause. That
+	// function contains the logic for determining whether a SINGLE document
+	// (Project) is a match for the Event. For a large number of Projects, such a
+	// query is wildly inefficient, since the function needs to be applied to
+	// every single document (Project). To compensate, we build PRELIMINARY match
+	// criteria based on matching source, type, and qualifiers. MongoDB sensibly
+	// applies these criteria FIRST, then iterates over the preliminary results
+	// only, applying the provided "where" function to each in turn.
+	preliminaryMatchCriteria := bson.M{
 		"source": event.Source,
 		"types": bson.M{
 			"$in": []string{event.Type, "*"},
 		},
 	}
-	if len(event.Labels) > 0 {
-		subscriptionMatchCriteria["labels"] = event.Labels
+	if len(event.Qualifiers) > 0 {
+		preliminaryMatchCriteria["qualifiers"] = event.Qualifiers
 	}
+	eventJSON, err := json.Marshal(
+		struct {
+			Source     string            `json:"source"`
+			Type       string            `json:"type"`
+			Qualifiers core.Qualifiers   `json:"qualifiers"`
+			Labels     map[string]string `json:"labels"`
+		}{
+			Source:     event.Source,
+			Type:       event.Type,
+			Qualifiers: event.Qualifiers,
+			Labels:     event.Labels,
+		},
+	)
+	// nolint: lll
+	where := fmt.Sprintf(`
+function() {
+	const project = this;
+	if (!project.spec.eventSubscriptions) {
+		return false;
+	}
+	const event = %s;
+	event.qualifiers = event.qualifiers || {};
+	event.labels = event.labels || {};
+	loop:
+	for (const subscription of project.spec.eventSubscriptions) {
```

Comment:
I wonder if there's a way to cover under test?  (Could be in a follow-up).  Maybe we try to test the JS directly (not via mongo or e2e) with pre-loaded data, to verify its logic is correct/intended?

Reply:
Good question. That question actually applies to all our queries (i.e. not just this one on account of the JS). Our unit tests for store code all simulate responses from the database, but they don't account at all for the possibility that a query is simply _wrong_.

What I think I'm banking on here is that once Brigade 2 reaches a point where we're using it to test itself, the new support for jobs with sidecars means that we can stand up a live MongoDB instance alongside our Go test container and actually start to put the store queries to the test.